src/Voron/Impl/LowLevelTransaction.cs

```diff
@@ -1383,20 +1112,17 @@ public void Rollback()
 
             ValidateReadOnlyPages();
 
-            foreach (var pageFromScratch in _transactionPages)
-            {
-                _env.ScratchBufferPool.Free(this, pageFromScratch.ScratchFileNumber, pageFromScratch.PositionInScratchBuffer, null);
-            }
-
-            foreach (var pageFromScratch in _unusedScratchPages)
-            {
-                _env.ScratchBufferPool.Free(this, pageFromScratch.ScratchFileNumber, pageFromScratch.PositionInScratchBuffer, null);
-            }
-
-            using (_env.PreventNewTransactions())
+            var rollbackPages = _env.WriteTransactionPool.ScratchPagesInUse;
+            
+            // we need to roll back all the changes we made here
+            _env.WriteTransactionPool.ScratchPagesInUse = _scratchBuffersSnapshotToRollbackTo.ToBuilder(); 
```

Comment:
Can you elaborate on this?

Reply:
Consider tx-10, which starts out with pages [1,2,3] in the scratch table.
We now modify pages 3 (again) and 4.

At the end of the transaction, we have the `_env.WriteTransactionPool.ScratchPagesInUse` with:
`[1,2,3 (new copy), 4]`

We need to roll it back to the previous state. So we clone the _previous_ transaction state, and dispose the new one.